File: reshapeTheMatrix.go
func matrixReshape(mat [][]int, r int, c int) [][]int {
    m, n := len(mat), len(mat[0])

    // Check if reshaping is possible
    if m*n != r*c {
        return mat
    }

    // Create a flattened list of elements
    flattened := make([]int, 0, m*n)
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            flattened = append(flattened, mat[i][j])
        }
    }

    // Construct the reshaped matrix
    reshaped := make([][]int, r)
    for i := 0; i < r; i++ {
        reshaped[i] = make([]int, c)
        for j := 0; j < c; j++ {
            reshaped[i][j] = flattened[i*c + j]
        }
    }
    return reshaped
}

File: reshapeTheMatrix.swift
class Solution {
    func matrixReshape(_ mat: [[Int]], _ r: Int, _ c: Int) -> [[Int]] {
        let m = mat.count
        let n = mat[0].count
        
        // Check if reshaping is possible
        if m * n != r * c {
            return mat
        }
        
        // Flatten the matrix into a single array
        var flattened = [Int]()
        for i in 0..<m {
            for j in 0..<n {
                flattened.append(mat[i][j])
            }
        }
        
        // Create the reshaped matrix
        var reshaped = [[Int]]()
        for i in 0..<r {
            var row = [Int]()
            for j in 0..<c {
                row.append(flattened[i * c + j])
            }
            reshaped.append(row)
        }
        
        return reshaped
    }
}
